agent/dns.go

```diff
@@ -1201,6 +1256,12 @@ RPC:
 		ttl, _ = d.GetTTLForService(out.Service)
 	}
 
+	// if the result came from cache, substract its age from the ttl
+	ttl -= age
```

Comment:
@Aestek & @pierresouchay 

I am not sure that the TTL calculations here are what is best/least-surprising.

One thing that initially caught me off guard was that this could potentially set a TTL lower than the remaining cache lifetime of the data.

For example, if you have a service ttl of 30s, a cache max age of 60s and your cached data is 20s old. This will set the TTL to 10s. At first glance you could ask why should a DNS client ask for the same data before the cache would have expired. If the cache hasn't expired and the request is made again it will just use the cached value the same as before. Therefore the TTL could have been something like `min(<remaining cache lifetime>, <service ttl - age>)` which would have resulted in a 40s TTL.

There are two reasons why that might not be best.

1. Automatic cache refreshing
2. The data might be fetched on behalf of the HTTP API with a different max age or requiring revalidation.

For prepared queries, there is no automatic cache refreshing so that doesn't matter right here.

So the question is whether it is better to have the service and node ttls be a max TTL for those resources or whether we should allow the cache lifetime to increase it. There are advantages and disadvantages to both. What are your thoughts?

Reply:
Are you suggesting `max(<remaining cache lifetime>, <service ttl - age>)` ? In your example `min(<remaining cache lifetime>, <service ttl - age>)` would still have resulted in a 20s TTL.
While I agree that its useless to have a TTL lower than the next cache refresh duration it seems weird to me to set a TTL higher than the one configured by the user.
When agent DNS cache the TTL settings can be lowered a lot since the requests would not hit so hard on the servers anymore. In our case we might juste use them for rate limiting with values between 1s and 5s.
Maybe it's best not to change the TTL depending on the cache age at all ?

Reply:
Yeah. Sorry, `max(<remaining cache lifetime>, <service ttl - age>`

However thinking about it some more it might be best the way you originally did it. It may result in more requests but they should be very easy to respond to from the cache.

Reply:
@Aestek @mkeeler As an DNS admin, I would like to enjoy the feature without having to handle the complexity of implementation. I have the feeling it should be simpler than that.

My feeling is:
 - `TTL` => stay unchanged
 - `max-age` = `max_stale` duration ideally (so, I don't care what your guys are doing of kind optimizations)

Thus, enabling cache would simply be:
 - `dns_config.enable_cache=true`